Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,364 @@
+  field_simp [r1, im_pos z]
+
+theorem r1_pos (z : ℍ) : 0 < r1 z := by
+  have H2 : 0 < (z.re ^ (2 : ℕ) + z.im ^ (2 : ℕ)) := by
+    apply_rules [pow_pos, add_pos_of_nonneg_of_pos, pow_two_nonneg, z.2]
+  exact div_pos (pow_pos z.im_pos 2) H2
+
+/-- This function is used to give an upper bound on Eisenstein series-/
+def r (z : ℍ) : ℝ := min (z.im) (Real.sqrt (r1 z))
+
+theorem r_pos (z : ℍ) : 0 < r z := by
+  simp only [r, lt_min_iff, im_pos, Real.sqrt_pos, r1_pos, and_self]
+
+lemma r_mul_pos_pos (k : ℕ) (z : ℍ) (n : ℝ) (hn : 0 < n) :
+    0 < (Complex.abs ((r z : ℂ) ^ (k : ℤ) * (n : ℂ)^ (k : ℤ))) := by
+  norm_cast
+  apply _root_.abs_pos.mpr (mul_ne_zero (pow_ne_zero k (ne_of_gt (r_pos z))) ?_)
+  simp only [Nat.cast_pow, ne_eq, pow_eq_zero_iff', Nat.cast_eq_zero, not_and, not_not] at *
+  intro _
+  linarith
+
+theorem r1_bound (z : ℍ) (δ ε : ℝ) (hε : 1 ≤ ε^2) :
+    (z.im ^ 2 ) / (z.re ^ 2 + z.im ^ 2) ≤ (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 := by
+  have H1 : (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 =
+        δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε^2 := by ring
+  have H4 : (δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε^2) * (z.re ^ 2 + z.im ^ 2)
+    - (z.im ^ 2) = (δ * (z.re ^ 2 + z.im ^ 2)+ ε * z.re)^2 + (ε^2 - 1)* (z.im)^2 := by ring
+  rw [H1, div_le_iff, ← sub_nonneg, H4]
+  · apply add_nonneg (pow_two_nonneg _) ?_
+    apply mul_nonneg
+    linarith
+    apply pow_two_nonneg
+  · apply_rules [add_pos_of_nonneg_of_pos, pow_two_nonneg, (pow_pos z.im_pos 2)]
+
+theorem auxbound1 (z : ℍ) (δ ε : ℝ) (hδ : 1 ≤ δ^2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs]
+  have H1 : (z : ℂ).im ≤
+    Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε) + (δ * z : ℂ).im * (δ * z : ℂ).im) := by
+    have h1 : (δ * z : ℂ).im * (δ * z : ℂ).im = δ^2 * (z : ℂ).im * (z : ℂ).im := by
+      simp only [mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul, add_zero]
+      ring
+    rw [Real.le_sqrt', h1 ]
+    nlinarith
+    exact z.2
+  simp only [UpperHalfPlane.coe_im, UpperHalfPlane.coe_re, AbsoluteValue.coe_mk, MulHom.coe_mk,
+    min_le_iff] at *
+  left
+  simp only [one_le_sq_iff_one_le_abs, mul_im, ofReal_re, coe_im, ofReal_im, coe_re, zero_mul,
+    add_zero, normSq_apply, add_re, mul_re, sub_zero, add_im] at *
+  exact H1
+
+theorem auxbound2 (z : ℍ) (δ ε : ℝ) (hε : 1 ≤ ε^2) : r z ≤ Complex.abs (δ * (z : ℂ) + ε) := by
+  rw [r, Complex.abs, min_le_iff]
+  have H1 : Real.sqrt (r1 z) ≤ Real.sqrt ((δ * (z : ℂ).re + ε) * (δ * (z : ℂ).re + ε ) +
+      δ * (z : ℂ).im * (δ * (z : ℂ).im)) := by
+    rw [r1, Real.sqrt_le_sqrt_iff, ← pow_two, ← pow_two]
+    apply r1_bound z δ ε hε
+    nlinarith
+  right
+  simp only [ne_eq, coe_re, coe_im, normSq_apply, AbsoluteValue.coe_mk, MulHom.coe_mk, add_re,
+    mul_re, ofReal_re, ofReal_im, zero_mul, sub_zero, int_cast_re, add_im, mul_im, add_zero,
+    int_cast_im] at *
+  exact H1
+
+--any suggestions as to where to put the lemmas below?
```

Comment:
Just ask `find_home`.

Reply:
Oh I didn't know about `find_home` neat!

Reply:
Note that you should use in a separate file importing the whole mathlib (see the documentation).